### PR TITLE
Add Jib detection to skaffold init

### PR DESCRIPTION
**Pull request by TadCordle**
Review after #2327.

This PR adds Jib detection to `skaffold init`. Skaffold walks the directory searching for dockerfiles, `build.gradle`s, and `pom.xml`s. If a `build.gradle` or a `pom.xml` is found, skaffold will attempt to run an internal Jib gradle task/maven goal to determine if Jib is configured on any of the projects (Jib v1.3.0 or higher is required for this to work).

If Jib is detected and `jib.to.image` is configured with an image found in the k8s manifest, the Jib builder is added to the skaffold config automatically.

If Jib is detected but `jib.to.image` is not configured to build a k8s image, the Jib project is added to the CLI prompt so the user can decide which image Jib should build, and the user's choice is added using the `-Djib.to.image=...` flag.

Issues left to solve:
- Configuring to build with Jib via CLI artifacts
- Gradle/Maven edge cases (config files that don't use the default names, unorthodox project structure, etc.)
- Generating Jib config automatically if it's missing?

Towards https://github.com/GoogleContainerTools/skaffold/issues/1264 (probably not completely fixed until there's some idea how to automatically apply Jib to maven/gradle projects).

### CODE REVIEW

File: pkg/skaffold/initializer/init.go [hunk 1]
```diff
@@ -221,7 +222,19 @@ func autoSelectBuilders(builderConfigs []InitBuilder, images []string) ([]builde
 	return pairs, builderConfigs, unresolvedImages
 }
 
-func detectBuilders(path string) ([]InitBuilder, error) {
+func detectBuilders(enableJibInit bool, path string) ([]InitBuilder, error) {
+	// TODO: Remove backwards compatibility if statement (not entire block)
+	if enableJibInit {
+		// Check for jib
+		if builders := jib.ValidateJibConfigFunc(path); builders != nil {
```
Comment (chanseokoh) [inline pkg/skaffold/initializer/init.go]: Isn't `ValidateJibConfig` more natural, given that `...Func` is like an ad-hoc variable only used by testing?

File: pkg/skaffold/jib/jib_init.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	return fmt.Sprintf("%s (%s)", j.BuilderName, j.FilePath)
+}
+
+// CreateArtifact creates an Artifact to be included in the generated Build Config
+func (j Jib) CreateArtifact(manifestImage string) *latest.Artifact {
+	workspace := filepath.Dir(j.FilePath)
+
+	a := &latest.Artifact{ImageName: j.Image}
+	if j.Image == "" {
+		a.ImageName = manifestImage
```
Comment (chanseokoh) [inline pkg/skaffold/jib/jib_init.go]: I forgot. So if a target image is configured in Jib (in `pom.xml` for example), then we can (or should) skip defining `image` in `skaffold.yaml`? This [repo example](https://github.com/GoogleContainerTools/skaffold/tree/master/examples) defines `image`s even if the images are configured in Jib.

File: pkg/skaffold/jib/jib_init.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+	// Parse Jib output. Multiple JSON strings may be printed for multi-project/multi-module setups.
+	matches := regexp.MustCompile(`BEGIN JIB JSON\r?\n({.*})`).FindAllSubmatch(stdout, -1)
+	if len(matches) == 0 {
+		return nil
+	}
+
+	results := make([]Jib, len(matches))
+	for i, match := range matches {
+		line := bytes.Replace(match[1], []byte(`\`), []byte(`\\`), -1)
```
Comment (chanseokoh) [inline pkg/skaffold/jib/jib_init.go]: What is this replacement for? Handling Windows path separators?

File: pkg/skaffold/jib/jib_init.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		parsedJSON := jibJSON{}
+		if err := json.Unmarshal(line, &parsedJSON); err != nil {
+			return nil
```
Comment (chanseokoh) [inline pkg/skaffold/jib/jib_init.go]: This should usually not happen, but may happen if we change the JSON structure? Should we log this or print a warning?

File: pkg/skaffold/jib/jib_init.go (hunk 2 above)
Comment (chanseokoh) [reply pkg/skaffold/jib/jib_init.go]: Never mind. I see `ImageName` is initialized at first.

File: pkg/skaffold/jib/jib_init.go (hunk 3 above)
Comment (TadCordle) [reply pkg/skaffold/jib/jib_init.go]: Yup, I can add a comment.

File: pkg/skaffold/initializer/init.go (hunk 1 above)
Comment (TadCordle) [reply pkg/skaffold/initializer/init.go]: I'm having trouble finding an example now, but I think I noticed this was a convention used elsewhere in the codebase.

File: pkg/skaffold/jib/jib_init.go (hunk 3 above)
Comment (chanseokoh) [reply pkg/skaffold/jib/jib_init.go]: It's unlikely path names on non-Windows will have `\`, so I think this is acceptable. BTW, how do you do a test for Windows?

**Pull request merged**